File: DSA/done/206 - reverseList.cpp
// https://leetcode.com/problems/reverse-linked-list/

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* rev = nullptr;
        while (head) {
            ListNode* tmp = head->next;
            head->next = rev;
            rev = head;
            head = tmp;
        }
        return rev;
    }
};

File: DSA/done/206 - reverseList.ts
// https://leetcode.com/problems/reverse-linked-list/

// Definition for singly-linked list.
class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

function reverseList(head: ListNode | null): ListNode | null {
    let rev: ListNode | null = null;
    while (head) {
        let tmp = head.next;
        head.next = rev;
        rev = head;
        head = tmp;
    }
    return rev;
};